Pare back default `index` option for the migration generator

prathamesh-sonpatki opened the pull request:
- [x] Update https://github.com/rails/rails/blob/master/activerecord/lib/active_record/migration/compatibility.rb to not create index by default unless specified for legacy migrations.
- Using `references` or `belongs_to` in migrations will always add index
  for the referenced column by default, without adding `index:true` option
  to generated migration file.
- Users can opt out of this by passing `index: false`.
- Fixes #18146

r? @sgrif cc @rafaelfranca

vipulnsward commented:
I believe, this will need a CHANGELOG.

prathamesh-sonpatki commented:
Yeah, I will add it also looking at failures from railties.

prathamesh-sonpatki commented on railties/test/generators/model_generator_test.rb:
```diff
@@ -345,26 +345,6 @@ def test_check_class_collision
-      end
-    end
-  end
-
```
I removed these tests. Should I add them for checking `index_is_not_added_for_references_association` ?

prathamesh-sonpatki commented:
Changelog added and build is green now.

sgrif commented:
r? @matthewd IIRC you were involved in the versioned migration changes. That's landed, correct? I'm not familiar enough with what's happening there to know if this only affects new migrations or if it affects old as well. If we haven't finished versioned migrations, and this is changing the behavior of existing migrations, I feel quite strongly that this should wait until 5.1

prathamesh-sonpatki commented:
@sgrif Yup, we have to update https://github.com/rails/rails/blob/master/activerecord/lib/active_record/migration/compatibility.rb to not create index by default unless specified for legacy migrations. I will work on it.

matthewd commented:
@prathamesh-sonpatki https://github.com/matthewd/rails/commit/f87c457ce3e1ddf446b75ce91e838cad2ca746fc may be helpful... I had the Actual Change, but hadn't gone back and chased it through the tests

prathamesh-sonpatki commented:
Thanks, I will check it.

prathamesh-sonpatki committed:
Pare back default `index` option for the migration generator

- Using `references` or `belongs_to` in migrations will always add index
  for the referenced column by default, without adding `index:true` option
  to generated migration file.
- Users can opt out of this by passing `index: false`.
- Legacy migrations won't be affected by this change. They will continue
  to run as they were before.
- Fixes #18146

prathamesh-sonpatki commented:
@matthewd Please review, took care of legacy migrations with your commit and added a test case for it.

matthewd commented:
:heart:
